---
title: "griph: Graph Inference of Population Heterogeneity"
author:
- name: Panagiotis Papasaikas
  email: panagiotis.papasaikas@fmi.ch  
- name: Michael Stadler
  email: michael.stadler@fmi.ch
- name: Atul Sethi
  email: atul.sethi@fmi.ch
date: "`r BiocStyle::doc_date()`"
package: "`r BiocStyle::pkg_ver('griph')`"
abstract: >
    griph identifies cell types in single cell RNA-seq datasets,
    allowing to control for unwanted sources of variance (e.g. cell cycle)
    that may confound the cell types.

    The most important function that combines all the steps of cell type identification
    from griph is griph_cluster. The vignette illustrates its use
    applied to several publicly available datasets.
output: 
    BiocStyle::html_document:
        keep_md: no
        toc_float: true
vignette: >
    %\VignetteIndexEntry{griph: Graph Inference of Population Heterogeneity}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
    \usepackage[utf8]{inputenc}
---


# Introduction and installation
Graph Inference of Population Heterogeneity (griph) is an R package for the analysis of
single cell RNA-sequencing data. It can be used to automatically identify different cell
types or states, even in the presence of confounding sources of variance such as
cell cycle stages or batch effects.

griph is currently available through https://github.com. It can be installed using:  
```{r install_github, eval=FALSE}
library(devtools)
install_git("git://github.com/ppapasaikas/griph.git", subdir = "griph")
```

In order for griph to work, some additional packages may have to be installed.
You'll need all the ones under "Needed". You can do without the ones under
"Optional", as they are only used in the vignette:  
```{r griph_dependencies, echo=FALSE}
get_deps <- function(path, which=c("Depends", "Imports", "Suggests")) {
    dcf <- read.dcf(file.path(path, "DESCRIPTION"))
    jj <- intersect(which, colnames(dcf))
    val <- unlist(strsplit(dcf[, jj], ","), use.names=FALSE)
    val <- gsub("\\s.*", "", trimws(val))
    val[val != "R"]
}
deps <- list(Needed=get_deps(system.file(package="griph"), c("Depends","Imports","LinkingTo")),
             Optional=get_deps(system.file(package="griph"), "Suggests"))
for(i in seq_along(deps))
    cat(names(deps)[i], ":\n\t", paste(deps[[i]], collapse=", "), "\n")
```

# Quickstart: A sample analysis
We will use human induced pluripotency cells from Tung et al. for this example (see 
details in the "Sample datasets" section). The dataset is included in the package, 
together with the iPS cell line labels:  
```{r tung_data}
M <- readRDS(system.file("extdata", "tung_UMIs_top10k.rds", package = "griph"))
dim(M) # genes by cells
trueLabel <- attr(M, "label2")
table(trueLabel)
```

Cell types can be identified using *griph_cluster*:  
```{r tung_griph_cluster, message=FALSE}
library(griph)
res <- griph_cluster(M, ClassAssignment = trueLabel, use.par = FALSE,
                     plot = TRUE, fsuffix = 'tung')
```
```{r tung_types}
table(res$MEMB)
```

The automatic generation of a 2D-embedding plot (`plot` argument) is
switched on here and will create a png (default) or pdf (`image.format`
and `fsuffix` arguments)
file in the current working directory:  
```{r showfile}
list.files("./", pattern = "tung")
```

The estimated cell graph (as an `r BiocStyle::CRANpkg("igraph")` object) and the coordinages
of the 2D-embedding (when using `plot=TRUE`) are returned by *griph_cluster*:  
```{r graphreturned}
summary(res$GRAO) # full graph
head(res$plotLVis) # 2D embedding
```


# Visualizing `r BiocStyle::Githubpkg("ppapasaikas/griph")` results
In the graph returned by *griph_cluster*, cells are represented by graph vertices.
All cells are contained in the full graph (`res$GRAO`), and a representatitve subset
of cells in the simplified graph (`res$plotGRAO`), which will be generated by *plotGraph*
and used for plotting.
The graph can be easily plotted again, for example to separately illustrate known and
predicted cell types:  
```{r knownVsPred, message=FALSE}
g.true <- plotGraph(res, fill.type = "true", line.type = "none")
res$plotGRAO <- g.true # store in res for re-use
g.pred <- plotGraph(res, fill.type = "predicted", line.type = "none")
```

In *plotGRAO*, weak edges are pruned, vertex attributes are added for visualization
of predicted and known class assignments (if given) and a subset of the vertices is sampled
if the graph exceeds the `maxG` argument. When `plotG` is set to `FALSE`,
the *plotGRAO* component in the return value of *griph_cluster* is `NULL` and
the simplified graph is generated later when necessary by *plotGraph*.

It is also possible to draw polygons around cells of each class:  
```{r tung_groupPolygons, message=FALSE}
g <- plotGraph(res, line.type = "none", mark.type = "true")
```

... or to collapse all the cells that are of the same class into a single vertex in
the graph (maybe useful for large graphs with many predicted classes):  
```{r tung_collapseSameClass, message=FALSE}
g <- plotGraph(res, collapse.type = "predicted")
```

Alternatively (typically for large graph with more than 1000 cells), results can
be visualized by applying a dimensionality reduction/projection technique such as
t-SNE to the affinity matrix returned by `r BiocStyle::Githubpkg("ppapasaikas/griph")`:  
```{r tung_tsne}
library(Rtsne)
out <- Rtsne(as.dist(1-res$DISTM), pca = FALSE,
             perplexity = 5, is_distance = TRUE)
plot(out$Y, xlab="tSNE-1", ylab="tSNE-2", col=res$MEMB,
     pch=as.numeric(trueLabel))
```

or more conveniently using the wrapper function provided by `griph`:  
```{r tung_tsne2, message=FALSE, echo=TRUE, results='hide'}
plotTsne(res, fill.type = "predicted", line.type = "none", mark.type = "true",
         perplexity = 7)
```

For even larger graphs (thousands of cells), a more efficient visualiztion uses
LargeVis (also see \url{https://github.com/elbamos/largeVis}, implementing the LargeVis
algorithm described by Tang et al. (2016) in \url{https://doi.org/10.1145/2872427.2883041})):  
```{r tung_largeVis, message=FALSE, echo=TRUE, results='hide'}
plotLVis(res, fill.type = "predicted", line.type = "none", mark.type = "true")
```

Finally, the plotting functions can also map numerical or categorical values to
the color of each cell. An example is to visualize the number of detected genes:  
```{r tung_detectedGenes, message=FALSE, echo=TRUE, results='hide'}
plotLVis(res, fill.type = colSums(M > 0), line.type = "none", mark.type = "true")
```

# Accessing the cell classification
*griph_cluster* identified `r length(unique(res$MEMB))`
cell types in the data, and the confusion matrix returned in the result summarizes
how they relate to the known cell types or states:  
```{r tung_eval}
res$ConfMatrix # confusion matrix
```

When comparing different classifications with each other, they may differ in granularity,
yet be consistent with each other. This can be seen for example in the confusion
matrix above where for example *NA190998* cells are found in multiple predicted classes. `r BiocStyle::Githubpkg("ppapasaikas/griph")`
calculates the classification error by assigning each identified class to the major known
class (for example *NA19098* for predicted class *2*) and only counting the 3 cells in *2*
as wrongly classified that are not *NA19098*:  
```{r tung_eval2}
res$miscl # misclassification error
```



# Sample datasets included in this package  

## Buettner et al., Nature Biotechnology 2015
Computational analysis of cell-to-cell heterogeneity in single-cell RNA-sequencing data reveals hidden subpopulations of cells. Buettner F, Natarajan KN, Casale FP, Proserpio V, Scialdone A, Theis FJ, Teichmann SA, Marioni JC, Stegle O. Nat Biotechnol. 2015 Feb;33(2):155-60. doi: 10.1038/nbt.3102.  

- Description: 288 mouse ES cells (Fluidigm C1), classified into three cell cycle stages
- PubMed: http://www.ncbi.nlm.nih.gov/pubmed/25599176
- DOI: https://doi.org/10.1038/nbt.3102
- Data: http://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-2805/
- Filtering: all cells, top 10,000 genes

```{r Buettner}
fname <- system.file("extdata", "buettner_top10k.rds", package = "griph")
M <- readRDS(fname)
dim(M)
label <- attr(M, "label")
table(label)
```


## Kolodziejczyk et al., Cell Stem Cell 2015
Single Cell RNA-Sequencing of Pluripotent States Unlocks Modular Transcriptional Variation. Kolodziejczyk AA, Kim JK, Tsang JC, Ilicic T, Henriksson J, Natarajan KN, Tuck AC, Gao X, Bühler M, Liu P, Marioni JC, Teichmann SA. Cell Stem Cell. 2015 Oct 1;17(4):471-85. doi: 10.1016/j.stem.2015.09.011.

- Description: 704 mouse ES cells (Fluidigm C1), three culture conditions: serum + LIF (lif), 2i + LIF (2i) and alternative 2i + LIF (a2i)
- PubMed: https://www.ncbi.nlm.nih.gov/pubmed/26431182  
- DOI: https://doi.org/10.1016/j.stem.2015.09.011  
- Data: http://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-2600/, http://www.ebi.ac.uk/teichmann-srv/espresso  
- Filtering: all cells, top 10,000 genes

```{r Kolodziejck}
fname <- system.file("extdata", "kolodziejck_top10k.rds", package = "griph")
M <- readRDS(fname)
dim(M)
label <- attr(M, "label")
table(label)
```

## Usoskin et al., Nat Neurosci. 2015
Unbiased classification of sensory neuron types by large-scale single-cell RNA sequencing. Usoskin D, Furlan A, Islam S, Abdo H, Lönnerberg P, Lou D, Hjerling-Leffler J, Haeggström J, Kharchenko O, Kharchenko PV, Linnarsson S, Ernfors P. Nat Neurosci. 2015 Jan;18(1):145-53. doi: 10.1038/nn.3881.

- Description: 799 mouse cells (622 neurons) from lumbar DRGs (custom protocol), different sensory neuron subtypes
- PubMed: https://www.ncbi.nlm.nih.gov/pubmed/25420068  
- DOI: https://doi.org/10.1038/nn.3881  
- Data: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE59739  
- Filtering: all cells, top 10,000 genes

```{r Usoskin}
fname <- system.file("extdata", "usoskin_top10k.rds", package = "griph")
M <- readRDS(fname)
dim(M)
label <- attr(M, "label")
table(label)
label2 <- attr(M, "label2")
table(label2)
label3 <- attr(M, "label3")
table(label3)
```

## Zeisel et al., Science 2015
Brain structure. Cell types in the mouse cortex and hippocampus revealed by single-cell RNA-seq. Zeisel A, Muñoz-Manchado AB, Codeluppi S, Lönnerberg P, La Manno G, Juréus A, Marques S, Munguba H, He L, Betsholtz C, Rolny C, Castelo-Branco G, Hjerling-Leffler J, Linnarsson S. Science. 2015 Mar 6;347(6226):1138-42. doi: 10.1126/science.aaa1934.

- Description: 3005 mouse brain cells (Fluidigm C1), different neuron subtypes
- PubMed: https://www.ncbi.nlm.nih.gov/pubmed/25700174  
- DOI: https://doi.org/10.1126/science.aaa1934  
- Data: http://linnarssonlab.org/cortex/, https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE60361  
- Filtering: all cells, top 10,000 genes

```{r Zeisel}
fname <- system.file("extdata", "zeisel_top10k.rds", package = "griph")
M <- readRDS(fname)
dim(M)
label <- attr(M, "label")
table(label)
label2 <- attr(M, "label2")
table(label2)
```


## Tung et al., Sci Rep 2017
Batch effects and the effective design of single-cell gene expression studies. Po-Yuan Tung, John D. Blischak, Chiaowen Joyce Hsiao, David A. Knowles, Jonathan E. Burnett, Jonathan K. Pritchard & Yoav Gilad. Scientific Reports 7, Article number: 39921 (2017). doi:10.1038/srep39921.

- Description: 864 human iPS cells (Fluidigm C1), three individuals, three C1 chip replicates for each individual  
- PubMed: https://www.ncbi.nlm.nih.gov/pubmed/28045081  
- DOI: http://dx.doi.org/10.1038/srep39921  
- Data: http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE77288, https://github.com/jdblischak/singleCellSeq  
- Filtering: all cells, top 10,000 genes

```{r Tung}
fname <- system.file("extdata", "tung_UMIs_top10k.rds", package = "griph")
M <- readRDS(fname)
dim(M)
label <- attr(M, "label")
table(label)
label2 <- attr(M, "label2")
table(label2)
```


# Session info {.unnumbered}
Here is the output of sessionInfo() on the system on which this document was compiled:
```{r sessionInfo, eval=TRUE, echo=FALSE}
sessionInfo()
```



```{r setup_documentation, echo=FALSE, message=FALSE, results='hide'}
#### Copy the created html to master root as index.html
#### and the md as README.md
  file.copy("griph.html","../../index.html",overwrite=TRUE)
#  file.copy("griph.md","../../README.md",overwrite=TRUE)
```

